### PR TITLE
KAFKA-6234: Increased timeout value for lowWatermark response to avoid test failing occasionally

### PR DESCRIPTION
**Pull request by soenkeliebau**
Increase timeout to fix flaky integration test testLogStartOffsetCheckpoint.

### CODE REVIEW

**Comment by asfgit** [comment]
SUCCESS 
 8083 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 7975 tests run, 5 skipped, 1 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8083 tests run, 5 skipped, 0 failed.
--none--

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: We should use the `get` method without a timeout. `waitUntilTrue` has its own timeout, so we don't need to have another one here.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, left one comment.

**Comment by soenkeliebau** [comment]
Thanks for the comment Ismael, that does indeed make much more sense. Have updated the PR.

**Comment by asfgit** [comment]
FAILURE
 7975 tests run, 5 skipped, 2 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8083 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8083 tests run, 5 skipped, 0 failed.
--none--

**Comment by soenkeliebau** [comment]
Not sure how that test failure came to pass, the exception should have been caught and retried. Apparently the test has transient failures for more than one scenario. Is this potentially a mixup of the scala and java versions of LeaderNotAvailableExceptions and thus the catch is ignored?

**Comment by ijuma** [comment]
The exception type is `ExecutionException` and `LeaderNotAvailableException` is the `cause`. The test needs to be fixed.

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (guozhangwang) [inline core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: Thanks for looking into this  @soenkeliebau . I agree with your reasoning on the JIRA. But I think as for the fix it is better to still set the timeout which is equal to the default value of `waitUntilTrue` (it is 15secs), because otherwise we may be blocked more than what is specified as the limit in `waitUntilTrue`, and then still catch the `TimeoutException` internally and return false.

cc @cmccabe

Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: Where do we catch the `TimeoutException` internally? Using a timeout in `get` is a bit of an anti-pattern, set the request timeout in `AdminClient` appropriately.

Comment (guozhangwang) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: 1. Regarding set the timeout: that's right, as long as we set the timeout config in `adminClient` that should be sufficient, that `future.get()` without parameter would still throw after configured timeout.
2. We should still catch `TimeoutException` in `future.get()` and return false;

Comment (soenkeliebau) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: I've added some code to unnest the LeaderNotAvailableException and rethrow if anything else is nested. Also catch the TimeoutException and return false.
As far as I can tell, timeouts are set to 20 in AdminClient and defaults to 15 for waitUntilTrue, so the Timeout _should_ never occur and we are fine - if it does, we handle it.
I'm not sure if in this case it wouldn't be better to explicitly state the timeouts though, to make the purpose of the test easier to understand without digging through default timeout values. Then again, I could just add a comment to that effect :)  I'll wait for your input before I push more commits..

Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: I don't think we want to catch the `TimeoutException` since it's longer than `waitUntilTrue` and it's a different failure mode than if `waitUntilTrue` times out because `LeaderNotAvailableException` happened for too long.

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: You can do this as:

```scala
case e: ExecutionException if e.getCause == LeaderNotAvailableException => false
```

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (soenkeliebau) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: Ok, I've removed the check for TimeoutException, since Guozhang hasn't weighed in again and he was advocationg for catching it.

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (soenkeliebau) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: I've refactored as per your suggestion, however the == didn't work as comparison for me since "LeaderNotAvailableException is not a value", so I changed it to isInstanceOf[].

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
retest this please

**Comment by soenkeliebau** [comment]
I had high hopes this time :)

**Comment by ijuma** [comment]
retest this please

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: The test is failing when the cause is `NotLeaderForPartitionException`. Is that the exception that you meant to use? Or do we need to catch both?

Comment (soenkeliebau) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: We've also seen it fail with LeaderNotAvailableException in the very beginning, so we probably need to catch both exceptions here. To be honest I never really questioned the original code that caught only LeaderNotAvailableException and simply fixed unnesting the exception.

Perhaps it would make sense to check for RetriableException instead?

Comment (soenkeliebau) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: I've changed the code to test for a retriable exception and as far as I can tell it does what it is supposed to do. This should also cover all other relevant cases that allow us to resubmit the delete request I think. For anything else it is ok to fail the test I think.

I've also rebased and squashed.

Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: I think we should stick to checking the two leader related exceptions since those are the only ones we expect to be thrown in this case.

Comment (soenkeliebau) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: I'm happy to change of course. I do wonder whether we might make the test too narrow though. If we encounter a retriable exception, retry and get the correct result, should we really fail the test?

**Comment by soenkeliebau** [comment]
Not sure what happened with the jdk7 test failure, it was this test case failing, but didn't look like an exception issue when I glanced at it, currently jenkins won't speak to me, so can't look in detail, will check back later.

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: Yes, if we are throwing a retriable exception that is unexpected, the test should fail. We can then check if we made a mistake in the test or the code.

**Comment by ijuma** [comment]
The latest error was because the low watermark was not 5. We should include the actual watermark in the error message. I'm going to do a separate PR to disable this test while we try to figure it out.

**Comment by soenkeliebau** [comment]
Looking through the log I was unsure whether we actually got a different low watermark or just ran into the timeout for waitUntilTrue because Exceptions kept us from ever getting to the actual check.

I'll include the returned value in the error message - we probably need to reset it before the second run, as it is still set to 5 from the first test - so if we run into the timeout due to exceptions the message would read "expected 5 but got 5" which might not help much :)
I'll look into setting an unitialized value so we can test for that and print a specific error message that allows to differentiate between "got a wrong value" and "got no value at all".

**Comment by ijuma** [comment]
Take a look at `TestUtils.computeUntilTrue` if `waitUntilTrue` makes it difficult to include the value in the error.

**Comment by soenkeliebau** [comment]
I've added some detail to the error message and reverted the catch back to just the two leader related exceptions.
It's not what I'd call beautiful code, there is probably some nice scala-y way of doing the check for two exceptions, but I couldn't come up with something nicer after googling a while tbh.
Let me know what you think of the extended error message.

**Comment by ijuma** [comment]
retest this please

**Comment by soenkeliebau** [comment]
retest this please

**Comment by soenkeliebau** [comment]
Hmm..it worked fine for two consecutive runs now, but a little while ago the JDK7 test consistently failed with the same code. Not sure if something else was changed (couldn't find anything, but only spent five minutes looking, so may very well have missed it) or if this was caused by something else entirely. I could never reproduce the test failure locally but it looked like ssl issues in the test log back when it failed, could this have been caused by load on the build server and some component taking too long to become ready?

**Comment by soenkeliebau** [comment]
retest this please

**Comment by soenkeliebau** [comment]
It seems to be back in working order for now. Unless someone has any idea of what might have caused the jdk7 tests to consistently fail back in November.
Are there any other comments on the PR as such? Otherwise I'd suggest merging and keeping an eye on whether the tests starts acting up again.

**Comment by soenkeliebau** [comment]
retest this please

**Comment by hachikuji** [comment]
retest this please

**Comment by soenkeliebau** [comment]
retest this please

**Comment by soenkeliebau** [comment]
retest this please

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: nit: I think it would be cleaner to use an `Option`

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: I wonder if it's still useful to pass a timeout to `get()` to avoid blocking longer than the timeout passed to `waitUntil`. We could use `JTestUtils.DEFAULT_MAX_WAIT_MS` or manually set a specific timeout.

nit: no need for parenthesis after `lowWatermark`.

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: nit: a bit more idiomatic to use string interpolation. If we make `lowWatermark` an option, then we can simplify this to just 
```scala
s"Expected low watermark of the partition to be 5 but got $lowWatermark"
```

**Review by hachikuji — COMMENTED** [review]
Sorry for the delay on review. Left a couple minor comments. Once addressed, I'm inclined to merge and see if the test case stabilizes.

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: Note that `requestTimeout` is 20 seconds, so I think `get` is OK.

**Comment by soenkeliebau** [comment]
Thanks for your comments @hachikuji 
I have hopefully addressed them in the last commit. I've changed your proposed log message a little, but I think it still looks cleaner than before.

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: It's a bit better to use `assertEquald(Some(5), lowWatermark)` so that you get a good error message.

Comment (soenkeliebau) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: Ok, I'll change that  once Jason has chimed in so that I can address his comments as well if he has any.

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: nit: Could we use `Some`? Same below.

**Review by hachikuji — APPROVED** [review]
Thanks, LGTM. Just had a minor nitpick.

File: core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala
Comment (soenkeliebau) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: I initially used Some(..) but subsequently learned that Some(null) is not None, however Option(null) is None and should be preferred so thought that it is better to use Option here (not that I'd expect lowWatermark to return null, but you never really know). 

But I really just read about this today, so this might be misguided :)

Comment (hachikuji) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: Yeah, that makes sense, but in this case `lowWatermark` is a `long` which cannot be null.

Comment (soenkeliebau) [reply core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala]: Fair point, I'll push a commit with this fixed and @ijuma 's concern addressed.

**Comment by hachikuji** [comment]
Thanks for the updates. I'll merge once the builds complete (assuming no problems).

**Pull request merged**
